docs/source/device.rst

```diff
@@ -0,0 +1,55 @@
+.. currentmodule:: torch
+
+.. _device-doc:
+
+torch.device
+===================================
+
+A :class:`torch.device` is an object representing the device on which a :class:`torch.Tensor` is
+or will be allocated.
+
+The :class:`torch.device` contains a device type ('cpu' or 'cuda') and optional device ordinal for the
+device type.  If the device ordinal is not present, this represents the current device for the device type;
+e.g. a :class:`torch.Tensor` constructed with device 'cuda' is equivalent to 'cuda:X' where X is the result of
+:func:`torch.cuda.get_device()`.
+
+A :class:`torch.device` can be constructed via a string, a string and device ordinal, or a device ordinal.
+
+
+Via a string:
+::
+
+    >>> torch.device('cuda:0')
+    device(device_type='cuda', device_index=0)
+
+    >>> torch.device('cpu')
+    device(device_type='cpu')
+
+    >>> torch.device('cuda')  # current cuda device
+    device(device_type='cuda')
+
+
+Via a string and device ordinal:
+
+::
+
+    >>> torch.device('cuda', 0)
+    device(device_type='cuda', device_index=0)
+
+    >>> torch.device('cpu', 0)
+    device(device_type='cpu', device_index=0)
+
+Via a device ordinal (device ordinals by themselves are treated as cuda devices).
+
+::
+
+  >>> torch.device(1)
+  device(device_type='cuda', device_index=1)
+
+.. note::
+   Methods which take a device will generally accept a (properly formatted) string
+   or integer device ordinal, e.g. the following are all equivalent:
+
+   >>> torch.randn((2,3), device=1)
```

Comment:
Would `device=0` turn into `cpu:0` or `cuda:0`?

Reply:
My bad, should have read more carefully. Looks like it turns into `cuda:0`.

Reply:
We should avoid the legacy way of specifying the device, can we remove it?

Reply:
The only issue there is it's kind of necessary for supporting legacy code, i.e. if you take a device parameter, it's safest to just wrap it in a `torch.device` and go from there.  So perhaps instead of interspersing the device ordinal examples with the others, we should move the device ordinal section to a separate legacy section (which can be linked from the migration guide).

Reply:
I think our policy for legacy code should be: it works, but we *never* show it in the docs, to avoid spreading it. If someone has been using this way then fine, but if they will look up the docs, then they should see the recommended ways.

Reply:
I think we should add it for one release and mark it as for legacy reasons.  My concern is people won't use the new stuff if it's not clear how to adapt the old code to it.